File: go-algorithm/leetcode/problem36.go
package leetcode

/*
36. Valid Sudoku
https://leetcode.com/problems/valid-sudoku/
*/
func isValidSudoku(board [][]byte) bool {
	rows, cols, boxes := make([]map[byte]bool, 9), make([]map[byte]bool, 9), make([]map[byte]bool, 9)
	for i := 0; i < 9; i++ {
		rows[i] = make(map[byte]bool)
		cols[i] = make(map[byte]bool)
		boxes[i] = make(map[byte]bool)
	}
	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			ch := board[i][j]
			if ch == '.' {
				continue
			}
			if rows[i][ch] {
				return false
			}
			rows[i][ch] = true
			if cols[j][ch] {
				return false
			}
			cols[j][ch] = true
			boxIndex := i/3 + (j/3)*3
			if boxes[boxIndex][ch] {
				return false
			}
			boxes[boxIndex][ch] = true
		}
	}
	return true
}

File: python-algorithm/leetcode/problem_36.py
"""36. Valid Sudoku
https://leetcode.com/problems/valid-sudoku/
"""
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for row in board:
            seen = set()
            for c in row:
                if c == '.':
                    continue
                if c in seen:
                    return False
                seen.add(c)
        for i in range(9):
            seen = set()
            for j in range(9):
                c = board[j][i]
                if c == '.':
                    continue
                if c in seen:
                    return False
                seen.add(c)
        for i in range(3):
            for j in range(3):
                seen = set()
                for x in range(3 * i, 3 * i + 3):
                    for y in range(3 * j, 3 * j + 3):
                        c = board[x][y]
                        if c == '.':
                            continue
                        if c in seen:
                            return False
                        seen.add(c)
        return True
